tlsn-core/src/session_header.rs

```diff
@@ -47,8 +79,54 @@ impl SessionHeader {
     pub fn handshake_summary(&self) -> &HandshakeSummary {
         &self.handshake_summary
     }
+}
+
+#[derive(Clone, Serialize, Default)]
+pub struct SessionHeaderMsg {
+    header: SessionHeader,
+    /// signature over `header`
+    signature: Option<Signature>,
+}
+
+impl SessionHeaderMsg {
+    pub fn new(header: &SessionHeader, signature: Option<Signature>) -> Self {
+        Self {
+            header: header.clone(),
+            signature,
+        }
+    }
+
+    /// Verifies the signature over the header against the public key. This is only called when we
+    /// know that `signature` is Some().
+    ///
+    /// Returns the verified header
+    fn verify(&self, pubkey: &PubKey) -> Result<SessionHeader, Error> {
+        let sig = match &self.signature {
+            Some(sig) => sig,
+            _ => return Err(Error::InternalError),
+        };
+
+        match (sig, pubkey) {
+            // signature and pubkey types must match
+            (Signature::P256(_), PubKey::P256(_)) => {
+                pubkey.verify(&self.header, &sig)?;
+            }
+        }
+
+        Ok(self.header.clone())
+    }
+
+    /// Returns the session header only if the signature is not present
+    pub fn get_header(&self) -> Result<SessionHeader, Error> {
+        match &self.signature {
+            Some(_) => Ok(self.header.clone()),
+            _ => {
+                return Err(Error::SignatureNotExpected);
+            }
+        }
+    }
 
-    pub fn signature(&self) -> Option<&[u8]> {
-        self.signature.as_deref()
+    pub fn signature(&self) -> Option<Signature> {
+        self.signature.clone()
```

Comment:
I took this liberty because if we use `as_ref()` then the caller would have to convert `Option<&Signature>` back to `Option<Signature>` when using it later. I thought this would pollute the API:

currently we do (in lib.rs line 80)

```rust
 let signature = msg.signature();
...
let session = NotarizedSession::new(1, header, signature, data);
```